src/quo2/components/record_audio/record_audio/view.cljs

```diff
@@ -0,0 +1,473 @@
+                                                         @recording?) (reset! ready-to-lock? moved-to-lock-button?)
+                                                        (and
+                                                         (or
+                                                          (and moved-to-record-button? @ready-to-delete?)
+                                                          (and moved-to-delete-button? @record-button-at-initial-position?))
+                                                         (not @ready-to-lock?)
+                                                         (not @ready-to-send?)
+                                                         @recording?) (reset! ready-to-delete? moved-to-delete-button?)
+                                                        (and
+                                                         (or
+                                                          (and moved-to-record-button? @ready-to-send?)
+                                                          (and moved-to-send-button? @record-button-at-initial-position?))
+                                                         (not @ready-to-lock?)
+                                                         (not @ready-to-delete?)
+                                                         @recording?) (reset! ready-to-send? moved-to-send-button?)))))
+                 :on-responder-release          (fn [^js e]
+                                                  (let [on-record-button? (touch-inside-area?
+                                                                           {:locationX (-> e .-nativeEvent.locationX)
+                                                                            :locationY (-> e .-nativeEvent.locationY)}
+                                                                           (if @reviewing-audio? record-button-area record-button-area-big)
+                                                                           false false false false)]
+                                                    (cond
+                                                      (and @reviewing-audio? on-record-button?)
+                                                      (reset! reviewing-audio? false)
+                                                      (and @ready-to-lock? (not @record-button-is-animating?))
+                                                      (do
+                                                        (reset! locked? true)
+                                                        (reset! ready-to-lock? false))
+                                                      (and (not @reviewing-audio?) on-record-button?)
+                                                      (do
+                                                        (when (>= @recording-length-ms 500) (reset! reviewing-audio? true))
+                                                        (reset! locked? false)
```

Comment:
I'm obviously not suggesting you to refactor the entire implementation, but it sure looks like managing all this state is fragile and difficult to get correct. It will be a nightmare for anyone having to change/evolve this namespace.

It would have been quite nice to see the separation of the state management from the actual view, something like a basic state machine could've been fruitful too. Obviously re-frame is out of question for this kind of performance requirements.

Reply:
Yeah, I know, the component is internally complex. It has so many "invisible" conditions based on touch starts, touch movements, touch releases, animations completion, and also different states. I tried the idea of a basic state machine, but had some problems with useEffect that is needed to trigger some animations (currently reacting to specific ratom updates on the different internal components). Also I think using state machine would also make useEffect execute to every state machine update and would make those functions even more complex, so decided to keep it as is at the time of developing.

Reply:
Understandable, it seems there's more to this problem than meets the eye.